### PR TITLE
do improvements to the scan all job

### LINKED ISSUE #9670: Simplify the scan all work flow and provide more progress metrics info for the scan all work
Avoid calling the API in the scan call job to trigger the scan actions for the artifacts.

### LINKED ISSUE #9705: Incorrect error context message for deleting immutable scanner registration
`"registration %s is not allowed to delete as it is immutable: scanner API: update"`

->

`"registration %s is not allowed to delete as it is immutable: scanner API: delete"`

### LINKED ISSUE #9722: The order of scanners in Interrogation Services often changes without rules
What can we help you?

image

### PR DESCRIPTION
**Pull request by steven-zou**
- update scan all job to avoid sending too many HTTP requests
- update scan controller to support scan options
- update the DB schema of the scan report to introduce requester
- introduce scan all metrics to report the overall progress of scan all job
- fix the status updating bug in the scan report
- enhance the admin job status updates
- add duplicate checking before triggering generic admin job
- update the DB scheme of the admin job

fix #9705
fix #9722
fix #9670

Signed-off-by: Steven Zou <szou@vmware.com>

### CODE REVIEW

File: src/common/dao/admin_job.go [hunk 1]
```diff
@@ -59,17 +59,27 @@ func DeleteAdminJob(id int64) error {
 }
 
 // UpdateAdminJobStatus ...
-func UpdateAdminJobStatus(id int64, status string) error {
+func UpdateAdminJobStatus(id int64, status string, statusCode uint16, revision int64) error {
```
Comment (wy65701436) [inline src/common/dao/admin_job.go]: please add a sample for the sql generted by beego.

File: src/core/api/admin_job.go [hunk 2]
```diff
@@ -241,6 +245,43 @@ func (aj *AJAPI) submit(ajr *models.AdminJobReq) {
...
+		if err != nil {
+			aj.SendInternalServerError(errors.Wrap(err, "AJAPI"))
+			return
+		}
+
+		if len(ajbs) > 0 {
+			jb := ajbs[0]
+
+			// Get the overall stats with the ID of the previous job
+			stats, err := scan.DefaultController.GetStats(fmt.Sprintf("%d", jb.ID))
```
Comment (wy65701436) [inline src/core/api/admin_job.go]: should add a condition to judge whether the job is `scan job`

File: src/core/api/scan_all.go [hunk 3]
```diff
@@ -112,3 +151,26 @@ func isScanEnabled() (bool, error) {
 
 	return len(l) > 0, nil
 }
+
+func getLatestScanAllJobIDByKind(kind string) (int64, error) {
```
Comment (wy65701436) [inline src/core/api/scan_all.go]: move it to admin_job API?

File: src/pkg/scan/api/scan/all_handler.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+func HandleCheckIn(checkIn string) {
+	if len(checkIn) == 0 {
+		// Nothing to handle, directly return
+		return
+	}
+
+	ck := &all.CheckInData{}
+	if err := ck.FromJSON([]byte(checkIn)); err != nil {
+		log.Error(errors.Wrap(err, "handle check in"))
+	}
```
Comment (ywk253100) [inline src/pkg/scan/api/scan/all_handler.go]: Return?

File: src/pkg/scan/all/job.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
+// Copyright 2018 The Harbor Authors
```
Comment (wy65701436) [inline src/pkg/scan/all/job.go]: update the copyright?

File: src/pkg/scan/api/scan/all_handler.go (hunk 4 above)
Comment (steven-zou) [reply src/pkg/scan/api/scan/all_handler.go]: No, loop every artifact. For the error ones, just logged.

File: src/core/api/scan_all.go [hunk 6]
```diff
@@ -97,6 +99,41 @@ func (sc *ScanAllAPI) List() {
...
+		sts, err = scan.DefaultController.GetStats(fmt.Sprintf("%d", id))
+		if err != nil {
+			sc.SendInternalServerError(errors.Wrap(err, "get metrics: scan all API"))
+			return
+		}
+	}
+
+	// Return empty
+	if sts == nil {
+		sts = &all.Stats{}
```
Comment (wy65701436) [inline src/core/api/scan_all.go]: return http.StatusNotFound?

File: src/pkg/scan/all/job.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+	// List all the repositories of registry
+	// TODO: REPLACE DAO WITH CORRESPONDING MANAGER OR CTL
+	repos, err := dao.GetRepositories()
+	if err != nil {
+		err = errors.Wrap(err, "list repositories : scan all job")
+		logger.Error(err)
+		return err
+	}
+	logger.Infof("Found %d repositories", len(repos))
```
Comment (wy65701436) [inline src/pkg/scan/all/job.go]: logger.Debugf()?

Comment (steven-zou) [reply src/pkg/scan/all/job.go]: No, use the `info` to output it to the job log

File: src/pkg/scan/all/job.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+		go func(repo *models.RepoRecord) {
+			defer func() {
+				// Return the token when process ending
+				tokens <- true
+			}()
+
+			logger.Infof("Scan artifacts under repository: %s", repo.Name)
+
+			repoClient, err := utils.NewRepositoryClientForJobservice(repo.Name, sa.registryURL, sa.secret, sa.tokenServiceEndpoint)
```
Comment (wy65701436) [inline src/pkg/scan/all/job.go]: can we change it to query the table of artifact?

File: src/pkg/scan/all/stats.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+// limitations under the License.
+
+package all
+
+// Stats provides the overall progress of the scan all process.
+type Stats struct {
+	Total uint `json:"total"`
+	// Status including `Success`, `Error` or `Stopped` will be counted as completed.
+	// This data may be influenced by job retrying
+	Completed uint          `json:"completed"`
```
Comment (wy65701436) [inline src/pkg/scan/all/stats.go]: `finish` is better to explain the meaning.

File: src/pkg/scan/report/base_manager.go [hunk 10]
```diff
@@ -237,3 +239,48 @@ func (bm *basicManager) DeleteByDigests(digests ...string) error {
...
+		Metrics: make(all.StatusMetrics),
+	}
+
+	for k, v := range m {
+		// Increase the total metrics
+		sts.Total += v
+
+		cc := job.Status(k).Code()
+		// Increase the completed metrics
+		if cc == -1 || cc == job.SuccessStatus.Code() {
```
Comment (wy65701436) [inline src/pkg/scan/report/base_manager.go]: add comments to explain `-1` and job.SuccessStatus.Code() is actually `Stopped or Error or  Success`.

File: src/core/api/scan_all.go (hunk 6 above)
Comment (steven-zou) [reply src/core/api/scan_all.go]: No need

File: src/pkg/scan/all/job.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+			if err != nil {
+				logger.Errorf("Failed to get tags for repo: %s, error: %v", repo.Name, err)
+				return
+			}
+
+			if len(al) > 0 {
+				// Check in the data
+				arts := make([]*v1.Artifact, 0)
+
+				for _, a := range al {
```
Comment (wy65701436) [inline src/pkg/scan/all/job.go]: why do not return artifact itself? what's the reason to convert it to v1. Artifact

Comment (steven-zou) [reply src/pkg/scan/all/job.go]: That's a TODO I have left there. @wy65701436

**Pull request merged**